core/src/avm1/value.rs

```diff
@@ -246,92 +246,53 @@ impl<'gc> Value<'gc> {
-                    return Ok(true);
-                }
-
-                Ok(false)
-            }
-            (Value::String(a), Value::String(b)) => Ok(*a == *b),
-            (Value::Bool(a), Value::Bool(b)) => Ok(a == b),
-            (Value::Object(a), Value::Object(b)) => Ok(Object::ptr_eq(*a, *b)),
-            (Value::Object(a), Value::Undefined | Value::Null) => Ok(Object::ptr_eq(
-                *a,
-                activation.context.avm1.global_object_cell(),
-            )),
-            (Value::Undefined | Value::Null, Value::Object(b)) => Ok(Object::ptr_eq(
-                *b,
-                activation.context.avm1.global_object_cell(),
-            )),
-            (Value::Undefined, Value::Null) => Ok(true),
-            (Value::Null, Value::Undefined) => Ok(true),
-            (Value::Number(_), Value::String(_)) => Ok(self.abstract_eq(
-                Value::Number(other.coerce_to_f64(activation)?),
-                activation,
-                true,
-            )?),
-            (Value::String(_), Value::Number(_)) => {
-                Ok(Value::Number(self.coerce_to_f64(activation)?)
-                    .abstract_eq(other, activation, true)?)
+                // PLAYER-SPECIFIC: NaN == NaN returns true in Flash Player 7+ AVM1, but returns false in Flash Player 6 and lower.
+                // We choose to return true.
+                a == b || (a.is_nan() && b.is_nan())
             }
-            (Value::Bool(_), _) => Ok(Value::Number(self.coerce_to_f64(activation)?)
-                .abstract_eq(other, activation, true)?),
-            (_, Value::Bool(_)) => Ok(self.abstract_eq(
-                Value::Number(other.coerce_to_f64(activation)?),
-                activation,
-                true,
-            )?),
-            (Value::String(_), Value::Object(_)) => {
-                let non_obj_other = other.to_primitive_num(activation)?;
-                if !non_obj_other.is_primitive() {
-                    return Ok(false);
-                }
 
-                Ok(self.abstract_eq(non_obj_other, activation, true)?)
+            // Bool-to-value-comparison: Coerce bool to 0/1 and compare.
+            (Value::Bool(bool), val) | (val, Value::Bool(bool)) => {
+                val.abstract_eq(Value::Number(bool as i64 as f64), activation)?
             }
-            (Value::Number(_), Value::Object(_)) => {
-                let non_obj_other = other.to_primitive_num(activation)?;
-                if !non_obj_other.is_primitive() {
-                    return Ok(false);
-                }
 
-                Ok(self.abstract_eq(non_obj_other, activation, true)?)
+            // Number-to-value comparison: Coerce value to f64 and compare.
+            // Note that "NaN" == NaN returns false.
+            (Value::Number(a), string @ Value::String(_))
+            | (string @ Value::String(_), Value::Number(a)) => {
+                a == string.coerce_to_f64(activation)?
```

Comment:
A more specific function can be used to coerce a string into a number, by capturing the `AvmString` and using a portion of `primitive_as_number`. However, it's just an optimization so I think it's not crucial for this PR.